helper/src/crypto.rs

```diff
@@ -0,0 +1,93 @@
+    100000000000000,     200000000000000,     300000000000000,     400000000000000,     500000000000000,     600000000000000,     700000000000000,     800000000000000,     900000000000000,
+    1000000000000000,    2000000000000000,    3000000000000000,    4000000000000000,    5000000000000000,    6000000000000000,    7000000000000000,    8000000000000000,    9000000000000000,
+    10000000000000000,   20000000000000000,   30000000000000000,   40000000000000000,   50000000000000000,   60000000000000000,   70000000000000000,   80000000000000000,   90000000000000000,
+    100000000000000000,  200000000000000000,  300000000000000000,  400000000000000000,  500000000000000000,  600000000000000000,  700000000000000000,  800000000000000000,  900000000000000000,
+    1000000000000000000, 2000000000000000000, 3000000000000000000, 4000000000000000000, 5000000000000000000, 6000000000000000000, 7000000000000000000, 8000000000000000000, 9000000000000000000,
+    10000000000000000000
+];
+
+/// Runtime initialized H generator.
+static H_PRECOMP: LazyLock<VartimeEdwardsPrecomputation> =
+    LazyLock::new(|| VartimeEdwardsPrecomputation::new([*H, ED25519_BASEPOINT_POINT]));
+
+/// Runtime initialized zero commitment lookup table
+///
+/// ASSUMPTION: This function assume that the [`ZERO_COMMITMENT_DECOMPOSED_AMOUNT`]
+/// table is sorted.
+pub static ZERO_COMMITMENT_LOOKUP_TABLE: LazyLock<[(u64, EdwardsPoint); 172]> =
+    LazyLock::new(|| {
+        let mut lookup_table: [(u64, EdwardsPoint); 172] = [(0_u64, ED25519_BASEPOINT_POINT); 172];
+
+        for (i, amount) in ZERO_COMMITMENT_DECOMPOSED_AMOUNT.into_iter().enumerate() {
+            lookup_table[i] = (
+                amount,
+                (ED25519_BASEPOINT_POINT + *H * Scalar::from(amount)),
+            );
+        }
+
+        lookup_table
+    });
+
+/// This function compute the zero commitment given a specific amount.
+///
+/// It will first attempt to lookup into the table of known Pre-RCT value.
+/// Compute it otherwise.
+pub fn compute_zero_commitments(amount: u64) -> EdwardsPoint {
+    // OPTIMIZATION: We first make an arithmetic check that the value lies between valid pre-computed amounts. This permit to avoid the lookup cost if possible
+    if amount > 10 && amount % 10 != 0 {
+        return H_PRECOMP.vartime_multiscalar_mul([Scalar::from(amount), Scalar::from(1_u8)]);
+    }
+
+    // Binary search in lookup table (O(log2(n)) instead of O(n))
+    let (mut start, mut end): (usize, usize) = (0, (ZERO_COMMITMENT_LOOKUP_TABLE.len() - 1));
+    while start <= end {
+        let mid = (start + end) / 2;
+        let lookup = ZERO_COMMITMENT_LOOKUP_TABLE[mid].0;
+        match amount {
+            a if a == lookup => {
+                return ZERO_COMMITMENT_LOOKUP_TABLE[mid].1;
+            }
+            a if a < lookup => {
+                end = mid - 1;
+            }
+            _ => {
+                start = mid + 1;
+            }
+        }
```

Comment:
I think `clippy::comparison_chain` wants this:
```suggestion
        match amount.cmp(&lookup) {
            Ordering::Equal => {
                return ZERO_COMMITMENT_LOOKUP_TABLE[mid].1;
            }
            Ordering::Less => {
                end = mid - 1;
            }
            Ordering::Greater => {
                start = mid + 1;
            }
        }
```